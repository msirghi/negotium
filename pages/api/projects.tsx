import type { NextApiRequest, NextApiResponse } from 'next';

const projects = [
  {
    id: 'uuid-1',
    name: 'Project 1',
  },
  {
    id: 'uuid-2',
    name: 'Project 2',
  },
  {
    id: 'uuid-3',
    name: 'Project 3',
  },
];

export default function handler(
  req: NextApiRequest,
  res: NextApiResponse<{ projects: typeof projects }>
) {
  res.status(200).json({ projects });
}
