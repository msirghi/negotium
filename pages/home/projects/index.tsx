import { useRouter } from 'next/router';
import { useEffect, useState } from 'react';
import { ProjectContainer } from '../../../src/components/project';
import Routes from '../../../src/common/config/routes';
import { Fade } from '@mui/material';

export default function Projects() {
  const router = useRouter();
  const [projectId, setProjectId] = useState<string>('');

  useEffect(() => {
    setProjectId('');
    if (router.query && router.query.id) {
      setProjectId(router.query.id as string);
      return;
    }
    router.push(Routes.inbox);
  }, [router.query]);

  if (!projectId) {
    return <div />;
  }

  return (
    <Fade in timeout={1000}>
      <div>
        <ProjectContainer projectId={projectId} />
      </div>
    </Fade>
  );
}
