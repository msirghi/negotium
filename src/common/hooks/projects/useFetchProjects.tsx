import { useQuery } from 'react-query';
import { projectsRequests } from '../../requests/projectsRequests';

export function useFetchProjects() {
  return useQuery('projectData', projectsRequests.fetchProjects);
}
