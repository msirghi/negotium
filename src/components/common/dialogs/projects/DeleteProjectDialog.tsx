import { StandardDialog } from '../standard/StandardDialog';
import { FC } from 'react';
import { DialogOptions } from '../standard/types';
import { Box } from '@mui/system';
import { makeStyles } from '@mui/styles';

type Props = {
  open: boolean;
  setOpen: (val: boolean) => void;
  onSubmit: () => void;
};

const useStyles = makeStyles({
  content: {
    textAlign: 'center',
    lineHeight: 1,
  },
});

export const DeleteProjectDialog: FC<Props> = ({ open, onSubmit, setOpen }) => {
  const options: DialogOptions = {
    dialogTitle: 'Delete project',
    cancelButtonLabel: 'Cancel',
    submitButtonLabel: 'Submit',
  };

  const classes = useStyles();

  return (
    <StandardDialog
      open={open}
      onSubmit={onSubmit}
      setOpen={setOpen}
      options={options}
    >
      <Box className={classes.content}>
        The project and all tasks associated with it will be deleted. <pre />{' '}
        Are you sure?
      </Box>
    </StandardDialog>
  );
};
