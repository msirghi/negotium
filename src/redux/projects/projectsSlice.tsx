import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { IProject } from '../../common/types/projects.types';

export interface ProjectsState {
  projects: IProject[];
}

const initialState: ProjectsState = {
  projects: [],
};

export const counterSlice = createSlice({
  name: 'projects',
  initialState,
  reducers: {
    setProjectsList: (state, action: PayloadAction<IProject[]>) => {
      state.projects = action.payload;
    },
    removeProjectFromList: (state, action: PayloadAction<string>) => {
      state.projects = state.projects.filter((p) => p.id !== action.payload);
    },
  },
});

// Action creators are generated for each case reducer function
export const { setProjectsList, removeProjectFromList } = counterSlice.actions;

export default counterSlice.reducer;
