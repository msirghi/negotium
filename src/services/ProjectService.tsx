import Requests from '../common/requests/request';
import { BASE_API_URL_V1 } from '../common/constants/constants';
import { Project } from '../common/types/projects.types';
import { Section, Task } from '../common/types/tasks.types';
import ServiceResultFactory from '../common/requests/serviceResultFactory';
import { HttpMethod } from '../common/constants/enums';
import { GetProjectResponse } from '../common/requests/types';

const addProject = (project: Omit<Project, 'id'>) => {
  return Requests.restApiCallWithBearer(`${BASE_API_URL_V1}/projects`, HttpMethod.POST, {
    ...project,
  });
};

const addProjectTask = (projectId: Project['id'], task: Omit<Task, 'id'>) => {
  return Requests.restApiCallWithBearer(`${BASE_API_URL_V1}/projects/${projectId}/tasks`, HttpMethod.POST, { ...task })
    .then(ServiceResultFactory.fromResponse)
    .catch(ServiceResultFactory.fromError);
};

const updateProjectTask = (projectId: Project['id'], task: Task) => {
  return Requests.restApiCallWithBearer(`${BASE_API_URL_V1}/projects/${projectId}/tasks/${task.id}`, HttpMethod.PATCH, { ...task })
    .then(ServiceResultFactory.fromResponse)
    .catch(ServiceResultFactory.fromError);
};

const updateProjectTaskDescription = (projectId: Project['id'], taskId: Task['id'], description: Task['description']) => {
  return Requests.restApiCallWithBearer(`${BASE_API_URL_V1}/projects/${projectId}/tasks/${taskId}`, HttpMethod.PATCH, { description })
    .then(ServiceResultFactory.fromResponse)
    .catch(ServiceResultFactory.fromError);
};

const getProjectById = (projectId: Project['id']) => {
  return Requests.restApiCallWithBearer(`${BASE_API_URL_V1}/projects/${projectId}`, HttpMethod.GET);
};

const updateProjectName = (projectId: Project['id'], name: Project['name']) => {
  return Requests.restApiCallWithBearer(`${BASE_API_URL_V1}/projects/${projectId}`, HttpMethod.PATCH, { name });
};

const updateProjectColor = (projectId: Project['id'], color: Project['color']) => {
  return Requests.restApiCallWithBearer(`${BASE_API_URL_V1}/projects/${projectId}`, HttpMethod.PATCH, { color });
};

const deleteProjectById = (projectId: Project['id']) => {
  return Requests.restApiCallWithBearer(`${BASE_API_URL_V1}/projects/${projectId}`, HttpMethod.DELETE);
};

const addProjectSection = (projectId: Project['id'], title: string) => {
  return Requests.restApiCallWithBearer(`${BASE_API_URL_V1}/projects/${projectId}/sections`, HttpMethod.POST, {
    title,
  })
    .then(ServiceResultFactory.fromResponse)
    .catch(ServiceResultFactory.fromError);
};

const getProjectSections = (projectId: string) => {
  return Requests.restApiCallWithBearer(`${BASE_API_URL_V1}/projects/${projectId}/sections`, HttpMethod.GET)
    .then(ServiceResultFactory.fromResponse)
    .then((res) => ServiceResultFactory.convertMongoIdToJSId(res.data))
    .catch(ServiceResultFactory.fromError);
};

const updateProjectSectionTitle = (projectId: Project['id'], sectionId: Section['id'], title: Section['title']) => {
  return Requests.restApiCallWithBearer(`${BASE_API_URL_V1}/projects/${projectId}/sections/${sectionId}`, HttpMethod.PATCH, { title })
    .then(ServiceResultFactory.fromResponse)
    .catch(ServiceResultFactory.fromError);
};

const deleteProjectSection = (projectId: Project['id'], sectionId: Section['id']) => {
  return Requests.restApiCallWithBearer(`${BASE_API_URL_V1}/projects/${projectId}/sections/${sectionId}`, HttpMethod.DELETE)
    .then(ServiceResultFactory.fromResponse)
    .catch(ServiceResultFactory.fromError);
};

const getProjects = () => {
  return Requests.restApiCallWithBearer(`${BASE_API_URL_V1}/projects`, HttpMethod.GET)
    .then(ServiceResultFactory.fromResponse)
    .then((res) => ServiceResultFactory.convertMongoIdToJSId(res.data))
    .catch(ServiceResultFactory.fromError);
};

const ProjectService = {
  addProject,
  getProjects,
  getProjectById,
  updateProjectName,
  deleteProjectById,
  addProjectTask,
  updateProjectTask,
  updateProjectTaskDescription,
  addProjectSection,
  getProjectSections,
  updateProjectSectionTitle,
  deleteProjectSection,
  updateProjectColor,
};
export default ProjectService;
