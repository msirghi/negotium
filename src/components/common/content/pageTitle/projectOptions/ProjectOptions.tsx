import { FC, useState, MouseEvent } from 'react';
import { IconButton, Menu, MenuItem } from '@mui/material';
import MoreVertIcon from '@mui/icons-material/MoreVert';
import { Row } from '../../../utilities/row/Row';
import DeleteOutlineOutlinedIcon from '@mui/icons-material/DeleteOutlineOutlined';
import { makeStyles } from '@mui/styles';
import { DeleteProjectDialog } from '../../../dialogs/projects/DeleteProjectDialog';
import { useRouter } from 'next/router';
import ProjectService from '../../../../../services/ProjectService';
import { useDispatch } from 'react-redux';
import { removeProjectFromList } from '../../../../../redux/projects/projectsSlice';
import { useSnackbar } from 'notistack';
import { SNACKBAR_POSITIONS } from '../../../../../common/constants/constants';
import DoneIcon from '@mui/icons-material/Done';

type Props = {};

const useStyles = makeStyles({
  menu: {
    marginTop: 35,
  },
  menuItem: {
    minWidth: 200,
  },
  menuItemTitle: {
    marginLeft: 10,
    fontSize: 14,
  },
});

export const ProjectOptions: FC<Props> = () => {
  const router = useRouter();
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const open = Boolean(anchorEl);
  const classes = useStyles();
  const [isDeleteDialogOpen, setDeleteDialogOpen] = useState(false);
  const dispatch = useDispatch();
  const { enqueueSnackbar } = useSnackbar();

  const onDeleteClick = () => {
    handleClose();
    setDeleteDialogOpen(true);
  };

  const onProjectDelete = async () => {
    const projectId = router.query.id as string;
    ProjectService.deleteProjectById(projectId);
    dispatch(removeProjectFromList(projectId));
    enqueueSnackbar('Project deleted', {
      anchorOrigin: SNACKBAR_POSITIONS.BOTTOM_CENTER,
    });
    await router.push('/home/inbox');
  };

  const handleClick = (event: MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  return (
    <div>
      <DeleteProjectDialog
        open={isDeleteDialogOpen}
        setOpen={setDeleteDialogOpen}
        onSubmit={onProjectDelete}
      />
      <IconButton onClick={handleClick}>
        <MoreVertIcon />
      </IconButton>
      <Menu
        className={classes.menu}
        anchorOrigin={{
          vertical: 'top',
          horizontal: 'right',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'right',
        }}
        id="basic-menu"
        anchorEl={anchorEl}
        open={open}
        onClose={handleClose}
        MenuListProps={{
          'aria-labelledby': 'basic-button',
        }}
      >
        <MenuItem onClick={onDeleteClick} className={classes.menuItem}>
          <Row alignVerticalCenter>
            <DoneIcon fontSize={'small'} />
            <span className={classes.menuItemTitle}>Show completed</span>
          </Row>
        </MenuItem>

        <MenuItem data-testid={'delete-item'} onClick={onDeleteClick} className={classes.menuItem}>
          <Row alignVerticalCenter>
            <DeleteOutlineOutlinedIcon color={'error'} fontSize={'small'} />
            <span className={classes.menuItemTitle}>Delete</span>
          </Row>
        </MenuItem>
      </Menu>
    </div>
  );
};
