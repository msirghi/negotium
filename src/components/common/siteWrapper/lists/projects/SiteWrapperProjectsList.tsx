import ListItemIcon from '@mui/material/ListItemIcon';
import ListItemText from '@mui/material/ListItemText';
import { List, ListItem } from '@mui/material';
import ReorderIcon from '@mui/icons-material/Reorder';
import MenuSkeleton from '../skeleton/MenuSkeleton';
import { SiteWrapperList } from '../';
import { ListItemTitle } from '../listItemTitle/ListItemTitle';
import { useFetchProjects } from '../../../../../common/hooks/projects/useFetchProjects';
import { useEffect, useState } from 'react';
import { ProjectDialog } from '../wrapper/projectDialog/ProjectDialog';
import { Project } from '../../../../../common/types/projects.types';
import ProjectService from '../../../../../services/ProjectService';
import { useRouter } from 'next/router';
import { makeStyles } from '@mui/styles';
import colors from '../../../../../common/styles/colors';
import { useDispatch, useSelector } from 'react-redux';
import { setProjectsList } from '../../../../../redux/projects/projectsSlice';
import { RootState } from '../../../../../redux/store';
import SmoothList from 'react-smooth-list';
import { ProjectListMoreItem } from './more/ProjectListMoreItem';
import { MAX_PROJECT_LIST_COUNT } from '../../../../../common/constants/constants';
import { useTranslation } from 'next-i18next';
import { Theme, useTheme } from '@mui/system';
import CircleIcon from '@mui/icons-material/Circle';

const useStyles = makeStyles({
  activeItem: {
    backgroundColor: (props: { theme: Theme }) => props.theme.palette.custom.menuIconBackground,
    // colors.greys['300'],
    color: colors.white,
    borderRadius: 10,
    transition: 'all .1s ease-in-out',
    '&:hover': {
      backgroundColor: (props: { theme: Theme }) => props.theme.palette.custom.menuIconBackground,
    },
  },
  activeIcon: {
    color: colors.white,
  },
});

export const SiteWrapperProjectsList = () => {
  const { isLoading, data, refetch } = useFetchProjects();
  const { t } = useTranslation('common');
  const theme = useTheme();
  const classes = useStyles({ theme });

  const [showAll, setShowAll] = useState(false);
  const [projects, setProjects] = useState<Project[]>([]);
  const [isProjectDialogOpen, setProjectDialogOpen] = useState(false);
  const projectsFromStore = useSelector((state: RootState) => state.projects.projects);

  const router = useRouter();
  const dispatch = useDispatch();

  useEffect(() => {
    if (data) {
      setProjects(data);
      dispatch(setProjectsList(data));
    }
  }, [data]);

  useEffect(() => {
    if (projectsFromStore) {
      setProjects(projectsFromStore);
    }
  }, [projectsFromStore]);

  useEffect(() => {
    refetch();
  }, [projectsFromStore]);

  const openDialog = () => setProjectDialogOpen(true);

  const onSubmit = async (name: Project['name'], color: Project['color']) => {
    await ProjectService.addProject({ name, color });
    await refetch();
  };

  const onProjectClick = (id: string) => {
    router.push(`/home/projects/${id}`);
  };

  const isProjectActive = (projectId: string) => {
    const { id } = router.query;
    return String(id) === String(projectId);
  };

  const toggleShowAll = () => setShowAll(!showAll);

  if (isLoading || !data) {
    return <MenuSkeleton />;
  }

  const handleProjectClick = (projectId: Project['id']) => {
    return () => onProjectClick(projectId);
  };

  return (
    <>
      <ProjectDialog
        dialogTitle={t('mainMenuList.addProject')}
        onSubmit={onSubmit}
        open={isProjectDialogOpen}
        setOpen={setProjectDialogOpen}
      />
      <SiteWrapperList title={'Projects'} options={{ addOptions: { onClick: openDialog } }}>
        <SmoothList>
          <List sx={{ padding: 1 }} data-testid={'projects-list'}>
            {projects.slice(0, showAll ? projects.length : MAX_PROJECT_LIST_COUNT).map((project) => {
              const isActive = isProjectActive(project.id);
              return (
                <ListItem
                  button
                  sx={{ borderRadius: 3 }}
                  className={isActive ? classes.activeItem : ''}
                  key={project.id}
                  onClick={handleProjectClick(project.id)}
                >
                  <ListItemIcon>
                    <CircleIcon style={{ color: project.color, fontSize: 13 }} className={isActive ? classes.activeIcon : ''} />
                  </ListItemIcon>
                  <ListItemText primary={<ListItemTitle title={project.name} />} />
                </ListItem>
              );
            })}
          </List>
          {projects.length > MAX_PROJECT_LIST_COUNT && <ProjectListMoreItem toggleShowAll={toggleShowAll} showAll={showAll} />}
        </SmoothList>
      </SiteWrapperList>
    </>
  );
};
