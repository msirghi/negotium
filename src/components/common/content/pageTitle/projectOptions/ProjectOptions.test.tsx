import { projectsMock } from '../../../../../common/tests/mockData/projects-mock';
import { MockReduxProvider } from '../../../../../common/tests/TestUtils';
import { SnackbarProvider } from 'notistack';
import { ProjectOptions } from './ProjectOptions';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { IconButton, Menu } from '@mui/material';
import { act } from '@testing-library/react';
import ProjectService from '../../../../../services/ProjectService';
import { DeleteProjectDialog } from '../../../dialogs/projects/DeleteProjectDialog';

const mockProjects = [...projectsMock];

jest.mock('next/router', () => ({
  useRouter: () => ({
    route: 'inbox',
    query: {
      id: mockProjects[0].id,
    },
    push: jest.fn(),
  }),
}));

describe('ProjectOptions', () => {
  const defaultProps = {
    onClick: jest.fn(),
  };
  const reduxStore = {};

  afterEach(() => {
    jest.clearAllMocks();
  });

  const renderComponent = () => {
    return (
      <MockReduxProvider reduxStore={reduxStore}>
        <SnackbarProvider>
          <ProjectOptions {...defaultProps} />
        </SnackbarProvider>
      </MockReduxProvider>
    );
  };

  it('should match the snapshot', () => {
    const tree = renderer.create(renderComponent());
    expect(tree).toMatchSnapshot();
  });

  it('should handle menu close', () => {
    const wrapper = mount(renderComponent());
    const menu = wrapper.find(Menu) as any;

    act(() => {
      menu.props().onClose();
    });
    wrapper.update();
    expect(menu.props().open).toBeFalsy();
  });

  it('should handle menu open', () => {
    const wrapper = mount(renderComponent());
    const icon = wrapper.find(IconButton).at(0);

    act(() => {
      icon.simulate('click');
    });
    wrapper.update();
    const menu = wrapper.find(Menu) as any;
    expect(menu.props().open).toBeTruthy();
  });

  it('should handle project delete', () => {
    const wrapper = mount(renderComponent());
    ProjectService.deleteProjectById = jest.fn();
    const dialog = wrapper.find(DeleteProjectDialog);

    act(() => {
      dialog.props().onSubmit();
    });
    expect(ProjectService.deleteProjectById).toBeCalled();
  });
});
