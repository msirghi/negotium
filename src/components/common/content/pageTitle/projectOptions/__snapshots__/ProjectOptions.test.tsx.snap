// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectOptions should match the snapshot 1`] = `
<div>
  <button
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="MoreVertIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
      />
    </svg>
  </button>
</div>
`;
